server/games/game_results.py

```diff
@@ -0,0 +1,136 @@
+@with_logger
+class GameResults(Mapping):
+    """
+    Collects all results from a single game. Allows to determine results for an
+    army and game as a whole. Supports a dict-like access to lists of results
+    for each army, but don't modify these.
+    """
+    def __init__(self):
+        Mapping.__init__(self)
+        self._back = {}
+
+    def __getitem__(self, key: int):
+        return self._back[key]
+
+    def __iter__(self):
+        return iter(self._back)
+
+    def __len__(self):
+        return len(self._back)
+
+    def add(self, result: GameResult):
+        army_results = self._back.setdefault(result.army, [])
+        army_results.append(result)
+
+    def is_mutually_agreed_draw(self, player_armies):
+        # Can't tell if we have no results
+        if not self:
+            return False
+        # Everyone has to agree to a mutual draw
+        for army in player_armies:
+            if army not in self:
+                continue
+            if any(r.outcome is not GameOutcome.MUTUAL_DRAW
+                   for r in self[army]):
+                return False
+        return True
+
+    def outcome(self, army: int) -> GameOutcome:
+        """
+        Determines what the game outcome was for a given army. Returns the
+        outcome all players agree on, excluding players that reported an
+        unknown outcome. Reports unknown outcome if players disagree.
+        """
+        if army not in self:
+            return GameOutcome.UNKNOWN
+
+        outcomes = set(r.outcome for r in self[army]
+                       if r.outcome is not GameOutcome.UNKNOWN)
+
+        if len(outcomes) == 1:
+            return outcomes.pop()
+        else:
+            return GameOutcome.UNKNOWN
+
+    def score(self, army: int, game_id=None):
+        """
+        Pick and return most frequently reported score for an army. If multiple
+        scores are most frequent, pick the largest one.
+        """
+        if army not in self:
+            return 0
+
+        scores = Counter(r.score for r in self[army])
+        if len(scores) == 1:
+            return scores.popitem()[0]
+
+        # FIXME - we pass game id just for logging
+        self._logger.info("Conflicting scores (%s) reported for game %s",
+                          scores, game_id)
+        score, _ = max(scores.items(), key=lambda kv: kv[::-1])
+        return score
+
+    def victory_only_score(self, army: int):
+        """
+        Calculate our own score depending *only* on victory.
+        """
+        if army not in self:
+            return 0
+
+        if any(r.outcome is GameOutcome.VICTORY for r in self[army]):
+            return 1
+        else:
+            return 0
+
+    # FIXME: Should we gather such methods in one class, or place them near
+    # corresponding types?
+    @classmethod
+    async def from_db(cls, database, game_id):
+        results = cls()
+        async with database.acquire() as conn:
+            rows = await conn.execute(
```

Comment:
Could rewrite this with sqlalchemy models

Reply:
Sure. I kept it to make it easier to verify that the code is equivalent, we can change that in a separate commit.